linkerd/app/core/src/control.rs

```diff
@@ -358,150 +188,148 @@ pub mod dns_resolve {
                                 } = inner.kind()
                                 {
                                     debug!("resolved does not exist");
-                                    this.current_targets.clear();
-                                    State::Resolved(
-                                        vec![Update::DoesNotExist].into_iter().collect(),
-                                        time::delay_until(time::Instant::from_std(
-                                            valid_until.clone(),
-                                        )),
-                                    )
-                                } else {
-                                    return Poll::Ready(Err(
-                                        dns::Error::ResolutionFailed(inner).into()
+                                    current_targets.clear();
+                                    pending_updates.push_back(Update::DoesNotExist);
+                                    ttl_expiration = time::delay_until(time::Instant::from_std(
+                                        valid_until.clone(),
                                     ));
+                                } else {
+                                    Err(dns::Error::ResolutionFailed(inner))?
                                 }
                             }
-                            Err(err) => return Poll::Ready(Err(err.into())),
                             Ok(result) => {
+                                resolve_ips = None;
                                 let new_targets = result
                                     .ips
                                     .into_iter()
                                     .map(|ip| Target {
-                                        addr: SocketAddr::from((ip, this.address.addr.port())),
-                                        server_name: this.address.identity.clone(),
+                                        addr: SocketAddr::from((ip, port)),
+                                        server_name: server_name.clone(),
                                     })
                                     .collect();
 
-                                let (adds, removes) =
-                                    Resolution::diff_targets(this.current_targets, &new_targets);
-                                let mut new_updates = VecDeque::default();
+                                let (adds, removes) = diff_targets(&current_targets, &new_targets);
                                 if !adds.is_empty() {
-                                    new_updates.push_back(Update::Add(adds))
+                                    pending_updates.push_back(Update::Add(adds))
                                 }
                                 if !removes.is_empty() {
-                                    new_updates.push_back(Update::Remove(removes))
+                                    pending_updates.push_back(Update::Remove(removes))
                                 }
```

Comment:
I think it _might_ be possible to simplify some of this logic now that we're using `async-stream`, by just `yield`ing these updates _here_ rather than queueing them. Since the stream will then resume _from_ the last `yield` when it's polled again, I think we might be able to avoid the queue entirely. 

It's fine if this ends up not being possible --- just a thought.